### PR TITLE
feature: Run from lenses in local terminal

### PR DESCRIPTION
**Pull request by tgodzik**
scala3-example-project-1666702645963

### CODE REVIEW

**Comment by tgodzik** [comment]
@kpodsiad what do you think? We can also generate lenses with run command and change the data that is being given, but that would not be that much gain and I think having the environment commands in metals server allows us for more flexibility.

**Review by kpodsiad — COMMENTED** [review]
It look promising, however I'll wait witch suggestions until we decide what should be done at server side.

File: src/scalaDebugger.ts [hunk 1]
```diff
@@ -75,7 +128,7 @@ class ScalaMainConfigProvider implements vscode.DebugConfigurationProvider {
         path: editor.document.uri.toString(true),
         runType: RunType.RunOrTestFile,
       };
-      return start(false, args).then(() => {
+      return start(debugConfiguration.noDebug, args).then(() => {
```
Comment (tgodzik) [inline src/scalaDebugger.ts:131]: Turns out that debugConfiguration contains the noDebug flag as a undeclared field. This is useful when running with F5 (Run without debugging)

File: src/scalaDebugger.ts [hunk 2]
```diff
@@ -31,10 +36,58 @@ export function initialize(outputChannel: vscode.OutputChannel): Disposable[] {
   ];
 }
 
+function isScalaRunMain(object: any): object is ScalaRunMain {
```
Comment (tgodzik) [inline src/scalaDebugger.ts:39]: I am not super sure about this one, but seems to do the job :sweat_smile:

**Review by kpodsiad — CHANGES_REQUESTED** [review]
I left a couple of suggestions

File: src/testExplorer/types.ts [hunk 3]
```diff
@@ -139,6 +139,33 @@ export interface TestSuiteRun {
...
+  targets: BuildTargetIdentifier[];
+}
+export interface OldScalaRunMain extends BaseScalaRunMain {
+  data: OldScalaMainData;
+}
+export interface ExtendedScalaRunMain extends BaseScalaRunMain {
+  data: ExtendedScalaMainData;
+}
+
+export type ScalaRunMain = OldScalaRunMain | ExtendedScalaRunMain;
```
[the remark this answers is not in the document]
Comment (kpodsiad) [reply src/testExplorer/types.ts:168]: Probably `ScalaTestSuites` and `ScalaTestSuiteSelection` should be moved also. Knowing how much you dislike typescipt, I could do those in separate PR ;)

[the remark this answers is not in the document]
Comment (tgodzik) [reply src/testExplorer/types.ts:168]: I don't dislike it, I am just not very good at it :sweat_smile:

File: src/scalaDebugger.ts [hunk 4]
```diff
@@ -57,12 +46,58 @@ export function initialize(outputChannel: vscode.OutputChannel): Disposable[] {
...
+  lenses: DebugDiscoveryParams | ScalaCodeLensesParams
+): lenses is ScalaRunMain {
+  return Object.prototype.hasOwnProperty.call(lenses, "data");
+}
+
+function isExtendedScalaRunMain(
+  runMain: DebugDiscoveryParams | ScalaCodeLensesParams
+): runMain is ExtendedScalaRunMain {
+  return isScalaRunMain(runMain) && runMain.data.shellCommand != undefined;
+}
```
[the remark this answers is not in the document]
Comment (tgodzik) [reply src/scalaDebugger.ts]: Thanks done!

File: src/scalaDebugger.ts [hunk 5]
```diff
@@ -57,12 +46,58 @@ export function initialize(outputChannel: vscode.OutputChannel): Disposable[] {
...
+    await tasks.executeTask(task);
+    return true;
+  }
+  return Promise.resolve(false);
+}
+
 export async function start(
   noDebug: boolean,
   debugParams: DebugDiscoveryParams | ScalaCodeLensesParams
 ): Promise<boolean> {
```
[the remark this answers is not in the document]
Comment (tgodzik) [reply src/scalaDebugger.ts]: Done!

File: src/scalaDebugger.ts [hunk 6]
```diff
@@ -57,12 +46,58 @@ export function initialize(outputChannel: vscode.OutputChannel): Disposable[] {
...
-  await commands.executeCommand("workbench.action.files.save");
+  if (noDebug && isExtendedScalaRunMain(debugParams)) {
+    return runMain(debugParams);
+  } else {
+    return debug(noDebug, debugParams);
+  }
+}
 
+export async function debug(
```
[the remark this answers is not in the document]
Comment (tgodzik) [reply src/scalaDebugger.ts]: Done!

File: src/scalaDebugger.ts [hunk 7]
```diff
@@ -6,35 +6,24 @@ import {
...
-  targets: BuildTargetIdentifier[];
-}
-
-export interface ScalaTestSuites {
-  data: FullyQualifiedClassName[];
-  dataKind: "scala-test-suites";
-  targets: BuildTargetIdentifier[];
-}
-
 export type ScalaCodeLensesParams = ScalaRunMain | ScalaTestSuites;
```
[the remark this answers is not in the document]
Comment (tgodzik) [reply src/scalaDebugger.ts]: Updated

File: src/testExplorer/types.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (tgodzik) [reply src/testExplorer/types.ts]: Ok, moved and everything seems to work fine. :sweat_smile:

**Review by kpodsiad — APPROVED** [review]
LGTM

**Pull request merged**